kvm-install-vm

```diff
@@ -434,7 +434,11 @@ function create_vm ()
 
     # cloud-init config: set hostname, remove cloud-init package,
     # and add ssh-key
-    cat > $USER_DATA << _EOF_
+    if [ ! -z "${USERDATA+x}" ]
+    then
+        . "${USERDATA}"
```

Comment:
Let's use `source` instead of `.` here.  I think it improves readability.

Now, if `${USERDATA}` is sourced, then users would have to remember to add `cat > $USER_DATA << _EOF_ ` at the top of their user data scripts, which is not intuitive.

How about:
```shell
if [ -n "${CUSTOM_USERDATA}" ]; then USERDATA=$(<${CUSTOM_USERDATA}); else ... fi
```

Reply:
I need to source my user password in the userdata file, that's why I source `${CUSTOM_USERDATA}`. In my custom userdata I do
```
# decrypt with openssl, e.g.:
# openssl aes-256-cbc -d -md sha512 -pbkdf2 -iter 1000 -salt -in /home/user01/file/enc/my-encrypted-password-hash
# then I source it
source /home/user01/file/enc/my-encrypted-password-hash
```
Let me finish testing the `openssl` decryption. Then I can share the exact example

Reply:
My complete userdata looks like:

```
#!/bin/bash                                                                                                                                                                                                                                                   
source <(openssl aes-256-cbc -d -md sha512 -pbkdf2 -iter 1000 -salt -in /home/user01/file/enc/my-encrypted-hash)
cat > $USER_DATA << _EOF_                                                                                                                                                                                                                                     
Content-Type: multipart/mixed; boundary="==BOUNDARY=="                                                                                                                                                                                                        
MIME-Version: 1.0                                                                                                                                                                                                                                             
--==BOUNDARY==                                                                                                                                                                                                                                                
Content-Type: text/cloud-config; charset="us-ascii"                                                                                                                                                                                                           
                                                                                                                                                                                                                                                              
#cloud-config                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                              
# Hostname management                                                                                                                                                                                                                                         
preserve_hostname: False                                                                                                                                                                                                                                      
hostname: ${VMNAME}                                                                                                                                                                                                                                           
fqdn: ${VMNAME}.${DNSDOMAIN}                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                              
# Users                                                                                                                                                                                                                                                       
users:                                                                                                                                                                                                                                                        
    - default                                                                                                                                                                                                                                                 
    - name: ${ADDITIONAL_USER}                                                                                                                                                                                                                                
      groups: ['${SUDOGROUP}']                                                                                                                                                                                                                                
      shell: /bin/bash                                                                                                                                                                                                                                        
      sudo: ALL=(ALL) ALL                                                                                                                                                                                                                            
      ssh-authorized-keys:                                                                                                                                                                                                                                    
        - ${KEY}                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                              
chpasswd:                                                                                                                                                                                                                                                     
  list: |                                                                                                                                                                                                                                                     
    root:${passwd_hash}                                                                                                                                                                                                                                       
    philipp:${passwd_hash}                                                                                                                                                                                                                                    
  expire: false                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                              
# Configure where output will go                                                                                                                                                                                                                              
output:                                                                                                                                                                                                                                                       
  all: ">> /var/log/cloud-init.log"                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                              
# configure interaction with ssh server                                                                                                                                                                                                                       
ssh_genkeytypes: ['ed25519', 'rsa']                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                              
# Install my public ssh key to the first user-defined user configured                                                                                                                                                                                         
# in cloud.cfg in the template (which is centos for CentOS cloud images)                                                                                                                                                                                      
ssh_authorized_keys:                                                                                                                                                                                                                                          
  - ${KEY}                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                              
timezone: ${TIMEZONE}                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                              
# Remove cloud-init when finished with it                                                                                                                                                                                                                     
runcmd:                                                                                                                                                                                                                                                       
  - ${NETRESTART}                                                                                                                                                                                                                                             
  - ${CLOUDINITDISABLE}                                                                                                                                                                                                                                       
_EOF_   
```

So while running `kvm-install-vm` I get prompted by `openssl` for my passphrase to decrypt my hash.

`my-encrypted-hash`
```
passwd_hash=\$6\$myhash..
```

Reply:
I see why you went this route. However, if we add an option to specify custom user data, then I would prefer that the provided file is a proper user data text file and not a shell script that needs to get sourced.  I would be OK with moving the password logic into the main script.

Instead of providing a hash, which is stored in plain text on your machine, what about coming up with a way where if you provided, say the `-P` flag, it will prompt your for a password at runtime (without echoing to the console) and then use `openssl` to create the hash for you. At that point, `"${PASSWD_HASH}"` would be available to the user data file.

Thoughts?

Reply:
We could even set defaults, like openssl arguments, in the main script and allow for overrides in `.kivrc`.

Reply:
## Option
There are multiple options:


### password-prompt
#### Description
- Ask for a password and generate it a SHA512 hash
- To generate the USER-DATA template use
  ```
  export ADDITIONAL_USER CLOUDINITDISABLE DNSDOMAIN KEY NETRESTART \
             passwd_hash SUDOGROUP TIMEZONE VMNAME
  USER_DATA=$(envsubst < "${CUSTOM_USERDATA}")
  ```
#### Pro:
- Automatic hash creation
- No heredoc inside custom `userdata` file
#### Con
- More maintainability
- More implementation
- Less flexibility:
  - You have to enter the password each time you create a vm. By using a heredoc the user has the flexibility to use e.g. `gpg-agent` to remember the password and create new vms without password prompt.
  - The user can't use another method to create the hash.
  - The user can't add more variables to Custom User Data
 - I get the error `ambigious redirect` when using `envsubst` together with the already implement `heredoc`. This has to write someone with more bash experience


### heredoc
#### Pro
- Flexibility
- Simpler to implement
- Less maintainability
#### Con
- Store heredoc in Custom User Data



### Plugin
#### Description
Allow sourcing a user plugin. The user plugin can contain a custom way to provide variables for use in the Custom User Data.
`.kivrc`
```
PLUGIN=plugin1, plugin2
```
`kvm-install-vm` (pseudo code)
```
for plugin in PLUGIN
   source plugin
```
- To generate the USER-DATA template use
  ```
  export ADDITIONAL_USER CLOUDINITDISABLE DNSDOMAIN KEY NETRESTART \
             passwd_hash SUDOGROUP TIMEZONE VMNAME
  USER_DATA=$(envsubst < "${CUSTOM_USERDATA}")
 
#### Pro
- Flexibility
#### Con
 - I get the error `ambigious redirect` when using `envsubst` together with the already implement `heredoc`. This has to write someone with more bash experience


### Combine Custom Plugin + Password
#### Pro
- Flexibility
- Easy to use for user
#### Con
- More maintainability
- More implementation

Reply:
I implemented now solution 3. That way the user's don't need to bother with the heredoc in the custom user data file. The password implementation can be added to the todo list. I hope this is OK now

Reply:
Implemented solution 4, using `-P` prompts for a password. I use python to generate the SHA-512 hash from the entered password and set it to the variable `PASSWD_HASH`, which can be used in the custom user data file.

Reply:
Hello @giovtorres 

is the Pull Request placed on hold or does it have a chance to be merged?

Kind Regards